### PR TITLE
Create edit reference image functionality

### PR DESCRIPTION
**Pull request by OlagunjuPhillips**

### CODE REVIEW

File: project/src/main/java/com/google/sps/data/ProductSearchLibrary.java [hunk 1]
```diff
@@ -275,17 +309,27 @@ private static void createProductSetItemAndAddToList(String productSetId, String
...
+        String formattedName =
+            ProductSearchClient.formatProductName(projectId, computeRegion, productId);
 
-            // Get the full path of the product.
-            String formattedName =
-                ProductSearchClient.formatProductName(projectId, computeRegion, productId);
+        // Delete a product.
+        client.deleteProduct(formattedName);
+        System.out.println("Product deleted.");
+      }
```
Comment (SirLegolot) [inline project/src/main/java/com/google/sps/data/ProductSearchLibrary.java]: I think you should either remove the try statement or create a catch so you can tell if deleting the product failed or not

File: project/src/main/java/com/google/sps/data/ProductSearchLibrary.java [hunk 2]
```diff
@@ -257,12 +256,47 @@ public static void removeProductFromProductSet(
...
+    String referenceImageId)
+    throws IOException {
+    try (ProductSearchClient client = ProductSearchClient.create()) {
+
+        // Get the full path of the reference image.
+        String formattedName =
+            ImageName.format(projectId, computeRegion, productId, referenceImageId);
+        // Delete the reference image.
+        client.deleteReferenceImage(formattedName);
+    }
```
Comment (SirLegolot) [inline project/src/main/java/com/google/sps/data/ProductSearchLibrary.java]: I think you should either remove the try statement or create a catch so you can tell if deleting the reference image failed or not

File: project/src/main/java/com/google/sps/servlets/BrowseServlet.java [hunk 3]
```diff
@@ -89,18 +89,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      //   productSearchIds.forEach(productId->imageSearchProducts.add(ServletLibrary.retrieveProductInfo(datastore, productId));
-      //   Set<ProductEntity> setProducts = new HashSet<>(products);
-      //   List<ProductEntity> newProducts = new ArrayList<>();
-      //   for (ProductEntity product : imageSearchProducts) {
-      //     if (setProducts.contains(product)) newProducts.add(product);
-      //   }
-      //   products = newProducts;
-      // }
+                                            
+    //   TODO: integrate once Phillips finishes product search.
```
Comment (SirLegolot) [inline project/src/main/java/com/google/sps/servlets/BrowseServlet.java]: Minor: indentation is a bit off

File: project/src/main/java/com/google/sps/servlets/ViewProductsServlet.java [hunk 4]
```diff
@@ -89,18 +89,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      //   productSearchIds.forEach(productId->imageSearchProducts.add(ServletLibrary.retrieveProductInfo(datastore, productId));
-      //   Set<ProductEntity> setProducts = new HashSet<>(products);
-      //   List<ProductEntity> newProducts = new ArrayList<>();
-      //   for (ProductEntity product : imageSearchProducts) {
-      //     if (setProducts.contains(product)) newProducts.add(product);
-      //   }
-      //   products = newProducts;
-      // }
+
+    //   TODO: integrate once Phillips finishes product search.
```
Comment (SirLegolot) [inline project/src/main/java/com/google/sps/servlets/ViewProductsServlet.java]: Minor: indentation is a bit off

**Review by SirLegolot — CHANGES_REQUESTED** [review]

**Review by SirLegolot — APPROVED** [review]

File: project/src/main/java/com/google/sps/data/ProductSearchLibrary.java [hunk 5]
```diff
@@ -234,8 +229,7 @@ public static void removeProductFromProductSet(
                 .setProductSearchParams(
                     ProductSearchParams.newBuilder()
                         .setProductSet(productSetPath)
-                        .addProductCategories(productCategory)
-                        .setFilter(filter))
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/ProductSearchLibrary.java:238]: Why is the filter being removed?

File: project/src/main/java/com/google/sps/data/ProductSearchLibrary.java [hunk 6]
```diff
@@ -275,17 +284,25 @@ private static void createProductSetItemAndAddToList(String productSetId, String
...
+        client.deleteProduct(formattedName);
+  }
+
+  public static String getProductId(String productName){
+
+      String[] productNameSplit = productName.split("/");
+
+      int productNameArrayLength = productNameSplit.length;
+
+      return productNameSplit[productNameArrayLength - 1];
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/ProductSearchLibrary.java]: You could skip a step here and just do `return productNameSplit[productNameSplit.length - 1];`

File: project/src/main/java/com/google/sps/servlets/BrowseServlet.java [hunk 7]
```diff
@@ -89,18 +89,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      //   productSearchIds.forEach(productId->imageSearchProducts.add(ServletLibrary.retrieveProductInfo(datastore, productId));
-      //   Set<ProductEntity> setProducts = new HashSet<>(products);
-      //   List<ProductEntity> newProducts = new ArrayList<>();
-      //   for (ProductEntity product : imageSearchProducts) {
-      //     if (setProducts.contains(product)) newProducts.add(product);
-      //   }
-      //   products = newProducts;
-      // }
+                                            
+      //   TODO: integrate once Phillips finishes product search.
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/BrowseServlet.java]: Can this TODO be removed now?

File: project/src/main/java/com/google/sps/servlets/ViewProductsServlet.java [hunk 8]
```diff
@@ -89,18 +89,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      //   List<ProductEntity> newProducts = new ArrayList<>();
-      //   for (ProductEntity product : imageSearchProducts) {
-      //     if (setProducts.contains(product)) newProducts.add(product);
-      //   }
-      //   products = newProducts;
-      // }
+
+      //   TODO: integrate once Phillips finishes product search.
+      if (searchInfo.getGcsUrl() != null) {
+        List <String> productSearchIds = ProductSearchLibrary.getSimilarProductsGcs("cloudberryAllProducts",
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/ViewProductsServlet.java]: You could define a constant for "cloudberryAllProducts" and then reference it wherever you need it (just to reduce risk of typos causing bugs).

File: project/src/main/java/com/google/sps/servlets/ViewProductsServlet.java [hunk 9]
```diff
@@ -89,18 +89,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      //   products = newProducts;
-      // }
+
+      //   TODO: integrate once Phillips finishes product search.
+      if (searchInfo.getGcsUrl() != null) {
+        List <String> productSearchIds = ProductSearchLibrary.getSimilarProductsGcs("cloudberryAllProducts", 
+                                            searchInfo.getProductCategory(), searchInfo.getGcsUrl());
+        List<ProductEntity> imageSearchProducts = new ArrayList<>();
+        productSearchIds.forEach(productId->imageSearchProducts.add(ServletLibrary.retrieveProductInfo(datastore, productId)));
+        Set<ProductEntity> setProducts = new HashSet<>(products);
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/ViewProductsServlet.java]: This is a little hard to follow with the 3 variables:
- `products`
- `setProducts`
- `newProducts`

It would be helpful to have more descriptive variable names. E.g., `imageSearchProducts` is pretty clear.

**Review by mariabajzek — CHANGES_REQUESTED** [review]

File: project/src/main/java/com/google/sps/data/ProductSearchLibrary.java (hunk 5 above)
Comment (OlagunjuPhillips) [reply project/src/main/java/com/google/sps/data/ProductSearchLibrary.java:238]: The filter is supposed to be used in the product search, and we really have no way of implementing the filters, so we removed them. We could have passed in an empty string for the filter, but that would mess with the values to be returned

File: project/src/main/java/com/google/sps/servlets/BrowseServlet.java [hunk 10]
```diff
@@ -90,18 +90,19 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     if (searchId != null) {
       SearchInfo searchInfo = ServletLibrary.retrieveSearchInfo(datastore, searchId);
                                             
-      //   TODO: integrate once Phillips finishes product search.
       if (searchInfo.getGcsUrl() != null) {
-        List <String> productSearchIds = ProductSearchLibrary.getSimilarProductsGcs("cloudberryAllProducts", 
-                                            searchInfo.getProductCategory(), changeGcsFormat(searchInfo.getGcsUrl()));
+        String generalProductSetId = "cloudberryAllProducts";
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/BrowseServlet.java:94]: Ah, I meant making this a global constant, but now I see that you're actually using it in separate files so that might be overkill. You can feel free to change it back if you would prefer.

My thought was just that when you have something like this stored in a String, your compiler won't notice if the string is different in different places, so you could end up with a silent failure to load your products and you wouldn't immediately know why. But since this is a relatively small codebase and all of you working on it know this is the name you're using, it's fine to just use the String here like you were doing.

**Review by mariabajzek — APPROVED** [review]
Looks good! Thanks for clarifying the set variable names. I just left one more comment but you're free to just merge this if you don't want to make that change.
